### ISSUE TITLE
Enhancement: no-return-await with complex `return` argument

### ISSUE THREAD

**Issue by daltones** [eslint/eslint#7594]
* **ESLint Version:** v3.10.0
* **Node Version:** v7.1.0
* **npm Version:** v3.10.9

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**
```json
{
  "root": true,
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module"
  },
  "rules": {
    "no-return-await": 1
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```js
async function foo1() {
  return (0, await bar());
}

async function foo2() {
  return (0, 1, await bar());
}

async function foo3() {
  return (baz() ? await bar() : 1);
}

async function foo4() {
  return (0 ? 1 : await bar());
}

async function foo5() {
  return (0 ? 1 : (2, 3, await bar()));
}

async function foo6() {
  return (await bar(), 1);
}

async function foo7() {
  return (await bar() ? 1 : 2);
}
```

**What did you expect to happen?**
`await` expressions in `foo1()`, `foo2()`, `foo3()`, `foo4()` and `foo5()` should raise “Redundant use of await on a return value” (`no-return-await` rule).

So, in cases where the `return` argument contains sequences or ternaries, it should search for redundant `await` expressions inside those expressions.

**What actually happened? Please include the actual, raw output from ESLint.**
No warnings at all.

**Comment by nzakas** [comment]
Hmm, I'm not sure about this. In general, we try not to get overly complex with the patterns we flag for rules like this, and I'm not sure if this is really the intent of the rule.

@ljharb do you have thoughts on this?

**Comment by ljharb** [comment]
My intuition is that someone using any of the examples in the OP actually does know what `await` is for, and understands its semantics inside `async function` - additionally, I doubt anyone genuinely uses the comma operator in hand-written code, and I don't think there's any value in linting against that.

I think that if there are more cases that can be caught, that's a great thing. However, these seem like they provide very low value with respect to the complexity detecting them adds.

**Comment by nzakas** [comment]
It looks like @daltones put together a PR for this already: https://github.com/eslint/eslint/pull/7595

@ljharb I'm inclined to agree with your comments, though would you object to inclusion given that a PR already exists?

**Comment by ljharb** [comment]
No, I don't object to inclusion.

**Comment by daltones** [comment]
Yeah, I was just trying to cover more cases. And the implementation is simple.

Of course the case for comma operators is very unlikely, but there's some that could have attention:

```js
// eg
return (condition ? something : await another);
return (something || await another);
```

**Comment by nzakas** [comment]
TSC Question: Should we accept this enhancement? (It just hasn't been looked at by anyone.)

**Comment by btmills** [comment]
Marking this as accepted per the 2016-12-01 TSC meeting.
